DokanNet/Dokan.cs

```diff
@@ -533,13 +197,52 @@ public static bool XAttrUpdate(DokanInstance dokanInstance, string filePath)
             /// <param name="isInSameDirectory">Indicates if the file or directory have the same parent directory.</param>
             /// <returns>true if notification succeeded.</returns>
             /// <remarks><see cref="DokanOptions.EnableNotificationAPI"/> must be set in the mount options for this to succeed.</remarks>
-            public static bool Rename(DokanInstance dokanInstance, string oldPath, string newPath, bool isDirectory, bool isInSameDirectory)
+            public static bool Rename(Dokan dokanInstance, string oldPath, string newPath, bool isDirectory, bool isInSameDirectory)
             {
                 return NativeMethods.DokanNotifyRename(dokanInstance.DokanHandle, oldPath,
                     newPath,
                     isDirectory,
                     isInSameDirectory);
             }
         }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            lock (_disposeLock)
+            {
+                if (!_disposed)
+                {
+                    if (disposing)
+                    {
+                        // dispose managed state (managed objects)
+                        DokanHandle?.Dispose();     // This calls DokanCloseHandle and waits for dismount
+                        DokanOptions?.Dispose();    // After that, it is safe to free unmanaged memory
+                        DokanOperations?.Dispose();
+                    }
+                    //  free unmanaged resources
+                    NativeMethods.DokanShutdown();

```

Comment:
This object is per mount instance but calling `DokanShutdown` on dispose will make all mount of the process to be shutdown.

Reply:
i tought there was a counter in the driver that release everything when it reaches zero, but maybe i'm mistaken

Reply:
No it does not work like that. This shutdown method cannot be called by instance cleanup code in this way. It can only be called once when everything Dokan related should be cleaned up.

Reply:
what would happen if two different processes calls that independetly ? is that safe?

Reply:
ok, i think next week i'll be able to work a little bit on this and propose something!